firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +124,105 @@ String getApplicationId() {
   String getName() {
     return firebaseApp.getName();
   }
+
+  /**
+   * Create firebase installation id for the {@link FirebaseApp} on FIS Servers and client side
+   * cache.
+   *
+   * <pre>
+   *     The workflow is:
+   *         check if cache empty or cache status is REGISTER_ERROR
+   *                                 |
+   *                           create random fid
+   *                                 |
+   *               update cache with cache status UNREGISTERED
+   *                                 |
+   *                           send http request to backend
+   *                                 |                  |
+   *  on success: set cache entry status to REGISTERED  |
+   *                                 |                  |
+   *                                 |         on failure: set cache entry status to REGISTER_ERROR
+   *                                 |                     |
+   *                                return          throw exception
+   *
+   *
+   *                      else if cached FID exists
+   *                             |            |
+   *     if cache status is UNREGISTERED      |
+   *                  |                   return cached FID
+   *     send http request to backend
+   *               |               |
+   *  on success: set cache        \
+   *  entry status to REGISTERED   \
+   *                |              \
+   *                |         on failure: set cache entry status to REGISTER_ERROR
+   *                |                     |
+   *              return          throw exception
+   * </pre>
+   */
+  @WorkerThread
+  private String createFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    FiidCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
+
+    if (cacheEntryValue == null
+        || cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.REGISTER_ERROR) {
+      String fid = Utils.createRandomFid();
+
+      boolean firstUpdateCacheResult =
+          localCache.insertOrUpdateCacheEntry(
+              FiidCacheEntryValue.create(fid, FiidCache.CacheStatus.UNREGISTERED, "", "", 0, 0));
```

Comment:
maybe use AutoValue and a Builder here?
Seems like a lot of parameters and it's unclear what , "'', "", 0, 0 does